File: app/src/main/java/com/atrainingtracker/trainingtracker/fragments/aftermath/EditDescriptionDialogFragment.java
package com.atrainingtracker.trainingtracker.fragments.aftermath;

import android.app.Dialog;
import android.content.ContentValues;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.atrainingtracker.R;
import com.atrainingtracker.trainingtracker.database.WorkoutSummariesDatabaseManager;

public class EditDescriptionDialogFragment extends DialogFragment {

    private static final String ARG_WORKOUT_ID = "workoutId";
    private static final String ARG_DESCRIPTION = "description";
    private static final String ARG_GOAL = "goal";
    private static final String ARG_METHOD = "method";

    private long workoutId;

    public interface OnDescriptionChangedListener {
        void onDescriptionChanged(long workoutId, String description, String goal, String method);
    }

    private OnDescriptionChangedListener mListener;

    public void setOnDescriptionChangedListener(OnDescriptionChangedListener listener) {
        mListener = listener;
    }

    public static EditDescriptionDialogFragment newInstance(long workoutId, String description, String goal, String method) {
        EditDescriptionDialogFragment fragment = new EditDescriptionDialogFragment();
        Bundle args = new Bundle();
        args.putLong(ARG_WORKOUT_ID, workoutId);
        args.putString(ARG_DESCRIPTION, description);
        args.putString(ARG_GOAL, goal);
        args.putString(ARG_METHOD, method);
        fragment.setArguments(args);
        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        if (getArguments() != null) {
            workoutId = getArguments().getLong(ARG_WORKOUT_ID);
        }

        AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_edit_description, null);

        final EditText etDescription = view.findViewById(R.id.edit_text_workout_description);
        final EditText etGoal = view.findViewById(R.id.edit_text_workout_goal);
        final EditText etMethod = view.findViewById(R.id.edit_text_workout_method);

        // Pre-fill the fields with existing data
        if (getArguments() != null) {
            etDescription.setText(getArguments().getString(ARG_DESCRIPTION));
            etGoal.setText(getArguments().getString(ARG_GOAL));
            etMethod.setText(getArguments().getString(ARG_METHOD));
        }

        builder.setView(view)
                .setTitle(R.string.title_edit_description) // Add this title string
                .setPositiveButton(R.string.save, (dialog, id) -> {
                    String newDescription = etDescription.getText().toString().trim();
                    String newGoal = etGoal.getText().toString().trim();
                    String newMethod = etMethod.getText().toString().trim();

                    if (mListener != null) {
                        mListener.onDescriptionChanged(workoutId, newDescription, newGoal, newMethod);
                    }
                })
                .setNegativeButton(R.string.cancel, (dialog, id) -> EditDescriptionDialogFragment.this.getDialog().cancel());

        return builder.create();
    }
}

File: app/src/main/java/com/atrainingtracker/trainingtracker/ui/components/workoutdescription/EditDescriptionDialogFragment.kt
package com.atrainingtracker.trainingtracker.ui.components.workoutdescription

import android.app.Dialog
import android.os.Bundle
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import com.atrainingtracker.R

class EditDescriptionDialogFragment : DialogFragment() {

    // The listener reports back the new data.
    var onDescriptionChanged: ((description: String, goal: String, method: String) -> Unit)? = null

    private val initialDescription: String by lazy { requireArguments().getString(ARG_DESCRIPTION, "") }
    private val initialGoal: String by lazy { requireArguments().getString(ARG_GOAL, "") }
    private val initialMethod: String by lazy { requireArguments().getString(ARG_METHOD, "") }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = requireActivity().layoutInflater.inflate(R.layout.dialog_edit_description, null)
        val etDescription = view.findViewById<EditText>(R.id.edit_text_workout_description)
        val etGoal = view.findViewById<EditText>(R.id.edit_text_workout_goal)
        val etMethod = view.findViewById<EditText>(R.id.edit_text_workout_method)

        etDescription.setText(initialDescription)
        etGoal.setText(initialGoal)
        etMethod.setText(initialMethod)

        return AlertDialog.Builder(requireContext())
            .setTitle(R.string.title_edit_description)
            .setView(view)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                // The dialog's ONLY job is to gather the text and invoke the listener.
                onDescriptionChanged?.invoke(
                    etDescription.text.toString(),
                    etGoal.text.toString(),
                    etMethod.text.toString()
                )
            }
            .setNegativeButton(android.R.string.cancel, null)
            .create()
    }

    companion object {
        private const val ARG_DESCRIPTION = "description"
        private const val ARG_GOAL = "goal"
        private const val ARG_METHOD = "method"

        @JvmStatic
        fun newInstance(description: String?, goal: String?, method: String?): EditDescriptionDialogFragment {
            return EditDescriptionDialogFragment().apply {
                arguments = bundleOf(
                    ARG_DESCRIPTION to description,
                    ARG_GOAL to goal,
                    ARG_METHOD to method
                )
            }
        }
    }
}